### PR TITLE
Directory restructure

### PR DESCRIPTION
**Pull request by alevy**
Restructuring directory according to https://groups.google.com/forum/#!topic/tock-dev/7d-FdWP6Zu0

This also rolls in changes to the build system that moves completely to a cargo-based build for the kernel and a separate, platform independent, build for userlevel example apps.

### CODE REVIEW

**Comment by alevy** [comment]
Any feedback on this?

**Comment by brghena** [comment]
Biggest issue: need to fix how make works.

**For the kernel:**
I consider it necessary that you can be in the root directory of the project, type `make`, and everything works. It seems pretty bad for new users to have to know to move into a specific directory in order to build the kernel.

**For the apps:**
I understand having to move into a specific example in order to build it. However, building an app should build libtock if not already done. Additionally, specifying an `ARCH=...` flag every time you want to build an app is no-go.

Basically, both of these are the same problem. We need a way to specify a default platform (which will have an architecture). How should we best do this? One way that comes to mind is to have a `configuration` file that just specifies the platform name.

**Comment by brghena** [comment]
Once we have make pinned down, the READMEs need to be updated. Currently half of each is out of date.

`tock/README.md`: Fix section `Building the kernel`
`tock/boards/storm/README.md`: Fix section `Programming user-level processes`

**Comment by brghena** [comment]
Can we move licenses to their own folder instead of having them muck up the root directory?

**Comment by ppannuto** [comment]
For the kernel build / make, is it necessary to work in every directory, or just the root? i.e. from the root, `$ make -C boards/storm` works fine, so we could add a trivial Makefile to make that work, but to work in every directory is going to require copying around a mostly empty makefile to every folder, which is just life with make.

**Comment by brghena** [comment]
What's going on with `arch/cortex-m*/syscalls.S`? It doesn't even look like it's used by anything anymore.

**Comment by brghena** [comment]
Do all the rust files _need_ to be in `<directory>/src/` rather than just in `<directory>/`?

**Comment by brghena** [comment]
The organization of `kernel/src/*` is a little unclear. What makes a file belong in `src/common/` rather than `src/`?

Additionally, the files in `kernel/src/support/` all seem to be ARM-specific. Should they go into `arch/` instead?

**Comment by brghena** [comment]
`tools/gdbinit_pca10028.jlink` and `tools/jdbserver_pca10028.sh` should be moved to `boards/nrf51dk/`

**Comment by brghena** [comment]
Does `rustfmt.toml` need to be in `tock/` or can it move to `tools/`?

**Comment by brghena** [comment]
With the new cargo build system, is an `elf` version of the kernel ever generated? I'd like to add in the capability for the build system to auto-generate `.lst` and `.Map` files for debugging.

**Comment by alevy** [comment]
Unfortunately, standalone comments aren't threaded (they are if they are inline to the code), so I'll just respond to everything in a mega comment:

> I consider it necessary that you can be in the root directory of the project, type make, and everything works. It seems pretty bad for new users to have to know to move into a specific directory in order to build the kernel.

We can add a Makefile with a target that just calls out to `make -C boards/storm`, potentially also compiling an example app (e.g. the blink app). We could let that be customizable at with a `TOCK_PLATFORM` variable.

> building an app should build libtock if not already done

Will do.

> ...We need a way to specify a default platform...

This is kind of getting at why I had suggested separate repos. Regardless, even with this, I think it makes sense to optimize for how a typical user will work with Tock, but that that should all be outside of this repo:

If you are a _user_ of Tock (e.g. you are writing apps), you will live in some external repo for your particular application. Your Makefile will define an `ARCH` variable to be whatever your test board is, by default.

If you are porting Tock to a board, you will also live in separate repo, that will look alot like the `boards/storm` directory, and it will depend on Tock through Cargo -- that's it. So your top level directory will have a Makefile that looks like `boards/storm/Makefile` (or whatever build system you want to use. Could be just cargo but there are some annoying arguments you have to pass).

I think it's fine to have a default of `ARCH=cortex-m4` in `userland/Makefile` since the default board should be Firestorm (well Imix soon, but same difference). Alternatively, a default of `ARCH=cortex-m0` would also work since M0 is forward compatible (i.e. same binaries will work on M4).

In general, though, I don't think we should optimize for the case of someone building end-to-end (applications + kernel) all in one go from the base directory.

> Can we move licenses to their own folder instead of having them muck up the root directory?

It's a bit unorthodox, and a handful of automated tools looks for LICENSE* files in the root directory.

> What's going on with arch/cortex-m*/syscalls.S?

Yeah, those should go

> Do all the rust files need to be in <directory>/src/ rather than just in <directory>/?

They don't have to, you can set some flag in `Cargo.toml` I think.

Again view this repo as a master repo, where each of major subdirectories is sort of an independent project. For cargo projects `src` is the default because you can also have `test`, `bench` and `examples` directories, all of which have their own use. We don't have any of these yet, but we very well may at some point soon. `src/` also helps separate source files from various ugly pluming files, e.g. `*.jlink`.

I don't feel strongly about it, but my intuition is: what reason to we have to go against the default?

> The organization of kernel/src/* is a little unclear. What makes a file belong in src/common/ rather than src/?

Agreed, but one thing at a time? For now I just literally turned those three crates into modules.

> Additionally, the files in kernel/src/support/ all seem to be ARM-specific. Should they go into arch/ instead?

Agreed. The only thing we really need (I think) in the kernel from `support` is `wfi` and `atomic`, so maybe those should just be part of the `Chip` trait (or maybe we add an `Arch` trait since there is starting to be a bit more functionality that is general across many chips).

Is this something we can defer until I take a larger pass on revising the arch crates anyway? (e.g. moving NVIC into the arch crates where it belongs).

> tools/gdbinit_pca10028.jlink and tools/jdbserver_pca10028.sh should be moved to boards/nrf51dk/

ack

> Does rustfmt.toml need to be in tock/ or can it move to tools/?

It helps. `rustfmt` looks for a `rustfmt.toml` anywhere in the directory tree up from where you are. So putting `rustfmt.toml` in the root directory means someone can run `cargo fmt` from any crate in the repo and the same style will apply.

> With the new cargo build system, is an elf version of the kernel ever generated? I'd like to add in the capability for the build system to auto-generate .lst and .Map files for debugging.

Yes, the result of the build is an ELF in `boards/$(BOARD)/target/target/$(crate-name)` (e.g. `boards/storm/target/target/storm`

(working on the annoying `target`. It's taken from the filename `target.json` which we still need to work on  a bit).
